### PR TITLE
Add FeatureFlag API

### LINKED ISSUE #8950: API to check if a material is disabled
### Is your feature request related to a problem?

Minecraft 1.19.3 added Feature Flag system, which allows some items to be disabled. I think there should be a way for server plugins to check if a material is disabled.

### Describe the solution you'd like.

An API to check if a bukkit Material is disabled or not in the specified world would be helpful. It needs a world since worlds can have different datapacks.

### Describe alternatives you've considered.

I'm currently using NMS codes to find out if a block is disabled or not (`net.minecraft.world.level.block.Block#isEnabled`)

### Other

_No response_

### PR DESCRIPTION
**Pull request by Machine-Maker**
Closes https://github.com/PaperMC/Paper/issues/8950

My initial thoughts on API for checking if a Material or EntityType is enabled with a `FeatureDependant` interface.

Adds tests to check on updates when this changes.

### CODE REVIEW

**Comment by maestro-denery** [comment]
How about using `@Deprecated` and `@APIStatus.ScheduledForRemoval()` for feature flags enabling updates? Also I think there should be a list of experimental feature flags available (all feature flags excluding the vanilla one) so people won't have to explicitly use fields which will be removed later to check whether a Material / EntityType is experimental or not.

**Comment by Machine-Maker** [comment]
> How about using @Deprecated and @APIStatus.ScheduledForRemoval() for feature flags enabling updates?

Deprecated seems inappropriate for this, and scheduledforremoval might work for UPDATE_1_20, but not bundle as we have no clue when that's gonna be removed. I think Experimental is the most accurate marker annotation.

File: patches/server/0968-Add-FeatureFlag-API.patch [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
++}
+diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
+index 781b72fd88149642c9fceaecfbfe7546273fb749..25726d007d8f0875bf86faa835717017edda0415 100644
+--- a/src/main/java/net/minecraft/server/Main.java
++++ b/src/main/java/net/minecraft/server/Main.java
+@@ -111,6 +111,7 @@ public class Main {
+                 JvmProfiler.INSTANCE.start(Environment.SERVER);
+             }
+ 
++            io.papermc.paper.world.flag.PaperFeatureFlagProviderImpl.INSTANCE.getClass(); // Paper - init FeatureFlagProvider
```
Comment (lynxplay) [inline patches/server/0968-Add-FeatureFlag-API.patch]: Is this the preferred way we want to inject server singletons into the API ?
I know this is a somewhat long standing issue between services, the server interface and this one.

**Review by kennytv — APPROVED** [review]

**Review by Owen1212055 — COMMENTED** [review]
Generally looks good, but think we should wait for the enum PR to allow us to have the "isEnabled" method be implemented manually to prevent the need for instance for the "FeatureElement" conversion thing.

**Comment by qixils** [comment]
Sorry, what exactly is blocking this? I don't see a "enum PR" from a glance, unless you mean the property API tagged "for: future" which I quite hope not... This is preventing me from being able to determine what items/entities are spawnable which is quite unfortunate for my use cases. I worked around it in the meantime with NMS but obviously this is not an ideal solution for cross-version compatibility among other issues. I'm hoping this could just be merged now (ideally to 1.19.4 while it's still supported ❤️) and improved later since it "generally looks good"?

**Comment by Machine-Maker** [comment]
The enum PR was the PR previous thought to be slated by upstream for 1.20. It replaces a ton of the enums and deprecates Material.

**Comment by qixils** [comment]
Ah, upstream, I can see the hesitation then. Unfortunate but I can live with NMS for the moment

**Comment by RedstoneFuture** [comment]
Would it be an option to extend the API with a method that outputs a permitted `EntityType[]` and `Material[]` for the corresponding feature-flag?

This could be helpful e.g. for tab-completions in which only the enabled blocks are to be displayed. Otherwise, you have to check all game objects individually each time.

**Comment by Machine-Maker** [comment]
Ok, revamped this and took over as much as possible from upstream's feature flag API. Added tests to ensure nothing is missed as well.

**Review by lynxplay — APPROVED** [review]

**Pull request merged**